src/platform/Darwin/DnssdHostNameRegistrar.cpp

```diff
@@ -0,0 +1,203 @@
+    uint16_t rdlen     = IsIPv4() ? sizeof(in_addr) : sizeof(in6_addr);
+    const void * rdata = IsIPv4() ? static_cast<const void *>(GetIPv4()) : static_cast<const void *>(GetIPv6());
+
+    DNSRecordRef dnsRecordRef;
+    return DNSServiceRegisterRecord(sdRef, &dnsRecordRef, kRegisterRecordFlags, mInterfaceId, hostname, rrtype, kDNSServiceClass_IN,
+                                    rdlen, rdata, 0, OnRegisterRecord, sdCtx);
+}
+
+bool InterfaceRegistrant::HasValidFlags(unsigned int flags)
+{
+    return !(flags & IFF_POINTOPOINT) && (flags & IFF_RUNNING) && (flags & IFF_MULTICAST);
+}
+
+bool InterfaceRegistrant::HasValidType(Inet::IPAddressType addressType, const sa_family_t addressFamily)
+{
+    bool useIPv4 = addressType == Inet::IPAddressType::kIPv4 || addressType == Inet::IPAddressType::kAny;
+    bool useIPv6 = addressType == Inet::IPAddressType::kIPv6 || addressType == Inet::IPAddressType::kAny;
+
+    return (useIPv6 && addressFamily == AF_INET6) || (useIPv4 && addressFamily == AF_INET);
+}
+
+bool InterfaceRegistrant::IsValidInterfaceId(uint32_t targetInterfaceId, unsigned int currentInterfaceId)
+{
+    return targetInterfaceId == kDNSServiceInterfaceIndexAny || targetInterfaceId == currentInterfaceId;
+}
+
+void InterfaceRegistrant::LogDetail() const
+{
+    char interfaceName[IFNAMSIZ] = {};
+    VerifyOrReturn(if_indextoname(mInterfaceId, interfaceName) != nullptr);
+
+    if (IsIPv4())
+    {
+        char addr[INET_ADDRSTRLEN] = {};
+        inet_ntop(AF_INET, GetIPv4(), addr, sizeof(addr));
+        ChipLogDetail(Discovery, "\tInterface: %s (%u) ipv4: %s", interfaceName, mInterfaceId, addr);
+    }
+    else
+    {
+        char addr[INET6_ADDRSTRLEN] = {};
+        inet_ntop(AF_INET6, GetIPv6(), addr, sizeof(addr));
+        ChipLogDetail(Discovery, "\tInterface: %s (%u) ipv6: %s", interfaceName, mInterfaceId, addr);
+    }
+}
+
+bool HostNameRegistrar::Init(const char * hostname, Inet::IPAddressType addressType, uint32_t interfaceId)
+{
+    mHostName = hostname;
+
+    // When the target interface is kDNSServiceInterfaceIndexLocalOnly, there is no need to map the hostname to
+    // an IP address.
+    VerifyOrReturnValue(interfaceId != kDNSServiceInterfaceIndexLocalOnly, true);
+
+    struct ifaddrs * ifap;
+    VerifyOrReturnValue(getifaddrs(&ifap) >= 0, false);
+
+    for (struct ifaddrs * ifa = ifap; ifa; ifa = ifa->ifa_next)
+    {
+        InterfaceRegistrant registrant;
+        if (registrant.Init(ifa, addressType, interfaceId))
+        {
+            mRegistry.push_back(registrant);
+        }
+    }
+
+    freeifaddrs(ifap);
+
+    return mRegistry.size() != 0;
+}
+
+CHIP_ERROR HostNameRegistrar::Register(RegisterContext * registerCtx)
+{
+    // If the target interface is kDNSServiceInterfaceIndexLocalOnly, there is no interfaces to register against
+    // the dns daemon.
+    if (mRegistry.size() == 0)
+    {
+        return registerCtx->Finalize();
```

Comment:
So if this returns error.... then the caller will also try to call `Finalize()` on registerCtx, no?  Seems like we should just ignore the return from `Finalize` and return CHIP_NO_ERROR here?